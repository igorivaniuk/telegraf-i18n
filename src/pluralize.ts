// https://developer.mozilla.org/en-US/docs/Mozilla/Localization/Localization_and_Plurals

const pluralRules: any = {
  english: (n: number) => (n !== 1 ? 1 : 0),
  french: (n: number) => (n > 1 ? 1 : 0),
  russian: (n: number) => {
    if (n % 10 === 1 && n % 100 !== 11) {
      return 0
    }
    return n % 10 >= 2 && n % 10 <= 4 && (n % 100 < 10 || n % 100 >= 20) ? 1 : 2
  },
  czech: (n: number) => {
    if (n === 1) {
      return 0
    }
    return n >= 2 && n <= 4 ? 1 : 2
  },
  polish: (n: number) => {
    if (n === 1) {
      return 0
    }
    return n % 10 >= 2 && n % 10 <= 4 && (n % 100 < 10 || n % 100 >= 20) ? 1 : 2
  },
  icelandic: (n: number) => (n % 10 !== 1 || n % 100 === 11 ? 1 : 0),
  chinese: () => 0,
  arabic: (n: number) => {
    if (n >= 0 && n < 3) {
      return n
    }
    if (n % 100 <= 10) {
      return 3
    }
    if (n >= 11 && n % 100 <= 99) {
      return 4
    }
    return 5
  }
}

const mapping: { [key: string]: string[] } = {
  english: ['da', 'de', 'en', 'es', 'fi', 'el', 'he', 'hu', 'it', 'nl', 'no', 'pt', 'sv', 'br'],
  chinese: ['fa', 'id', 'ja', 'ko', 'lo', 'ms', 'th', 'tr', 'zh', 'jp'],
  french: ['fr', 'tl', 'pt-br'],
  russian: ['hr', 'ru', 'uk', 'uz'],
  czech: ['cs', 'sk'],
  icelandic: ['is'],
  polish: ['pl'],
  arabic: ['ar']
}

export function pluralize(languageCode: string) {
  const language = languageCode.split('-')[0].toLowerCase()
  let key = Object.keys(mapping).find(key => mapping[key].includes(language))
  if (!key) {
    console.warn(`i18n::Pluralize: Unsupported language ${languageCode}`)
    key = 'english'
  }
  return (count: number, ...forms: string[]) =>
    `${count} ${forms[pluralRules[key as string](count)]}`
}

export function plural(languageCode: string) {
  const language = languageCode.split('-')[0].toLowerCase()
  let key = Object.keys(mapping).find(key => mapping[key].includes(language))
  if (!key) {
    console.warn(`i18n::Pluralize: Unsupported language ${languageCode}`)
    key = 'english'
  }
  return (count: number, ...forms: string[]) => `${forms[pluralRules[key as string](count)]}`
}
