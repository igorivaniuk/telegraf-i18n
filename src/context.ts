import { plural, pluralize } from './pluralize'

export interface Config {
  directory?: string
  useSession?: boolean
  sessionName: string
  allowMissing?: boolean
  defaultLanguage: string
}

export class I18nContext {
  defaultContext: object
  shortLanguageCode?: string

  constructor(
    private repository: any,
    public languageCode: string,
    templateData: object,
    public config: Config
  ) {
    this.defaultContext = templateData
    this.config = config
    this.locale(languageCode || this.config.defaultLanguage)
  }

  locale(languageCode?: string) {
    if (!languageCode) {
      return this.languageCode
    }

    const code = languageCode.toLowerCase()
    const shortCode = code.split('-')[0]

    if (!this.repository[code] && !this.repository[shortCode]) {
      this.languageCode = this.config.defaultLanguage
      this.shortLanguageCode = this.languageCode.split('-')[0]
      return
    }

    this.languageCode = code
    this.shortLanguageCode = shortCode
  }

  getTemplate(languageCode: string, resourceKey = '') {
    return resourceKey
      .split('.')
      .reduce((acc, key) => acc && acc[key], this.repository[languageCode])
  }

  t(resourceKey: string, context?: object) {
    let template =
      this.getTemplate(this.languageCode, resourceKey) ||
      this.getTemplate(this.shortLanguageCode!, resourceKey)
    if (!template) {
      if (!this.config.allowMissing) {
        throw new Error(`telegraf-i18n: '${this.languageCode}.${resourceKey}' not found`)
      }
      template = () => resourceKey
    }
    return template(
      Object.assign(
        {
          pluralize: pluralize(this.shortLanguageCode!),
          plural: plural(this.shortLanguageCode!)
        },
        this.defaultContext,
        context
      )
    )
  }
}
