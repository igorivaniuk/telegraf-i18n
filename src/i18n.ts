import * as fs from 'fs'
import * as yaml from 'js-yaml'
import * as path from 'path'
import { Config, I18nContext } from './context'

const compile = require('compile-template')

export class I18n {
  config: Config
  repository: { [key: string]: any } = {}

  constructor(config: Config) {
    this.config = Object.assign(
      {
        defaultLanguage: 'en',
        sessionName: 'session',
        allowMissing: true
      },
      config
    )
    if (this.config.directory) {
      this.loadLocales(this.config.directory)
    }
  }

  loadLocales(directory: string) {
    if (!fs.existsSync(directory)) {
      throw new Error(`Locales directory '${directory}' not found`)
    }
    const files = fs.readdirSync(directory)
    files.forEach(fileName => {
      const extension = path.extname(fileName)
      const languageCode = path.basename(fileName, extension).toLowerCase()
      if (extension === '.yaml' || extension === '.yml') {
        const data = fs.readFileSync(path.resolve(directory, fileName), 'utf8')
        this.loadLocale(languageCode, yaml.safeLoad(data))
      } else if (extension === '.json') {
        this.loadLocale(languageCode, require(path.resolve(directory, fileName)))
      }
    })
  }

  loadLocale(languageCode: string, i18Data: object) {
    const language = languageCode.toLowerCase()
    this.repository[language] = Object.assign(
      {},
      this.repository[language],
      compileTemplates(i18Data)
    )
  }

  resetLocale(languageCode?: string) {
    if (languageCode) {
      delete this.repository[languageCode.toLowerCase()]
    } else {
      this.repository = {}
    }
  }

  middleware() {
    return (ctx: { i18n: I18nContext; [key: string]: any }, next: () => Promise<any>) => {
      const session = this.config.useSession && ctx[this.config.sessionName]
      const languageCode =
        (session && session.__language_code) || (ctx.from && ctx.from.language_code)
      ctx.i18n = this.createContext(languageCode, {
        from: ctx.from,
        chat: ctx.chat
      })
      return next().then(rs => {
        if (session) {
          session.__language_code = ctx.i18n.locale()
        }
        return rs
      })
    }
  }

  createContext(languageCode: string, templateData: { [key: string]: any }) {
    return new I18nContext(this.repository, languageCode, templateData, this.config)
  }

  t(languageCode: string, resourceKey: string, templateData: object) {
    return this.createContext(languageCode, templateData).t(resourceKey)
  }
}

function compileTemplates(root: any) {
  Object.keys(root).forEach(key => {
    if (!root[key]) {
      return
    }
    if (Array.isArray(root[key])) {
      root[key] = root[key].map(compileTemplates)
    }
    if (typeof root[key] === 'object') {
      root[key] = compileTemplates(root[key])
    }
    if (typeof root[key] === 'string') {
      root[key] = compile(root[key])
    }
  })
  return root
}
