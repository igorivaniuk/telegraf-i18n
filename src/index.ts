import { ContextMessageUpdate } from 'telegraf'
import { ExtraReplyMessage } from 'telegraf/typings/telegram-types'
import { I18nContext } from './context'
import { I18n } from './i18n'
import { pluralize } from './pluralize'

function match(resourceKey: string, templateData: object) {
  return (text: string, ctx: { i18n: I18nContext }) =>
    text && ctx && ctx.i18n && text === ctx.i18n.t(resourceKey, templateData) ? [text] : null
}

function reply(resourceKey: string, extra: ExtraReplyMessage) {
  return ({ reply, i18n }: ContextMessageUpdate & { i18n: I18nContext }) =>
    reply(i18n.t(resourceKey), extra)
}

export { I18n, reply, match, pluralize, I18nContext }
