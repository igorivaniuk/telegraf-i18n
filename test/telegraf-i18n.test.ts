/**
 * Dummy test
 */
import * as path from 'path'
import { I18n } from '../src'

let templateData = { from: { id: 1234 }, chat: {} }

describe('I18n', () => {
  let i18n: I18n
  beforeAll(() => {
    i18n = new I18n({
      defaultLanguage: 'ru',
      allowMissing: true,
      directory: path.resolve(__dirname, 'locales'),
      sessionName: 'session'
    })
  })

  it('create ', () => {
    expect(i18n).toBeInstanceOf(I18n)
    expect(i18n.repository).toMatchSnapshot()
  })

  it('t ru', () => {
    expect(i18n.t('ru', 'command.start', templateData)).toBe('Привет 1234\n')
    expect(i18n.t('ru', 'button.settings', templateData)).toBe('Настройки')
    expect(i18n.t('ru', 'button.help', templateData)).toBe('Помощь')
  })

  it('t en', () => {
    expect(i18n.t('en', 'command.start', templateData)).toBe('Hello 1234\n')
    expect(i18n.t('en', 'button.settings', templateData)).toBe('Settings')
    expect(i18n.t('en', 'button.help', templateData)).toBe('Help')
  })

  it('t default lang', () => {
    expect(i18n.t('ru', 'command.start', templateData)).toBe('Привет 1234\n')
    expect(i18n.t('ru', 'button.settings', templateData)).toBe('Настройки')
    expect(i18n.t('ru', 'button.help', templateData)).toBe('Помощь')
  })

  it('plural en', () => {
    expect(
      i18n.t('en', 'plural', {
        ...templateData,
        count: 1
      })
    ).toBe('Found 1 user\n')

    expect(
      i18n.t('en', 'plural', {
        ...templateData,
        count: 2
      })
    ).toBe('Found 2 users\n')
  })

  it('plural ru', () => {
    expect(
      i18n.t('ru', 'plural', {
        ...templateData,
        count: 1
      })
    ).toBe('Найден 1 пользователь\n')

    expect(
      i18n.t('ru', 'plural', {
        ...templateData,
        count: 2
      })
    ).toBe('Найдено 2 пользователя\n')

    expect(
      i18n.t('ru', 'plural', {
        ...templateData,
        count: 3
      })
    ).toBe('Найдено 3 пользователя\n')
  })

  it('plural default lang', () => {
    expect(
      i18n.t('kkk', 'plural', {
        ...templateData,
        count: 1
      })
    ).toBe('Найден 1 пользователь\n')

    expect(
      i18n.t('kkk', 'plural', {
        ...templateData,
        count: 2
      })
    ).toBe('Найдено 2 пользователя\n')

    expect(
      i18n.t('kkk', 'plural', {
        ...templateData,
        count: 3
      })
    ).toBe('Найдено 3 пользователя\n')
  })

  it('resetLocale', () => {
    i18n.resetLocale('ru')
    expect(i18n.repository).toMatchSnapshot()

    i18n.resetLocale()
    expect(i18n.repository).toMatchSnapshot()
  })
})
