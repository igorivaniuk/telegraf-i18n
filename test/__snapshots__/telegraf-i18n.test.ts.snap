// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`I18n create  1`] = `
Object {
  "en": Object {
    "button": Object {
      "help": [Function],
      "settings": [Function],
    },
    "command": Object {
      "start": [Function],
    },
    "plural": [Function],
  },
  "ru": Object {
    "button": Object {
      "help": [Function],
      "settings": [Function],
    },
    "command": Object {
      "start": [Function],
    },
    "plural": [Function],
  },
}
`;

exports[`I18n resetLocale 1`] = `
Object {
  "en": Object {
    "button": Object {
      "help": [Function],
      "settings": [Function],
    },
    "command": Object {
      "start": [Function],
    },
    "plural": [Function],
  },
}
`;

exports[`I18n resetLocale 2`] = `Object {}`;
